pkg/liqo-controller-manager/namespaceMap-controller/ensure_remote_namespaces.go

```diff
@@ -0,0 +1,203 @@
+
+	mapsv1alpha1 "github.com/liqotech/liqo/apis/virtualKubelet/v1alpha1"
+	liqoconst "github.com/liqotech/liqo/pkg/consts"
+	liqoutils "github.com/liqotech/liqo/pkg/utils"
+)
+
+// This constants are used to compose the annotation that is inserted on remote Namespace at creation time.
+const (
+	liqoSuffix                      = "liqo.io"
+	remoteNamespaceAnnotationPrefix = "remote-namespace"
+	remoteNamespaceAnnotationValue  = "This Namespace has been created through Liqo offloading mechanism"
+)
+
+// This function gets the clusted-id of the local cluster.
+func (r *NamespaceMapReconciler) checkLocalClusterID() error {
+	if r.LocalClusterID == "" {
+		clusterID, err := liqoutils.GetClusterID(r.Client)
+		if err != nil || clusterID == "" {
+			return err
+		}
+		r.LocalClusterID = clusterID
+	}
+	return nil
+}
+
+// This function creates a remote Namespace inside the remote cluster, if it doesn't exist yet.
+// The right client to use is chosen by means of NamespaceMap's cluster-id.
+func (r *NamespaceMapReconciler) createRemoteNamespace(clusterID, remoteName string) error {
+	if err := r.checkRemoteClientPresence(clusterID); err != nil {
+		return err
+	}
+	if err := r.checkLocalClusterID(); err != nil {
+		return err
+	}
+
+	// This annotation is used to recognize the remote namespaces that have been created by this controller.
+	annotationKey := fmt.Sprintf("%s-%s/%s", remoteNamespaceAnnotationPrefix, r.LocalClusterID, liqoSuffix)
+	remoteNamespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: remoteName,
+			Annotations: map[string]string{
+				annotationKey: remoteNamespaceAnnotationValue,
+			},
+		},
+	}
+
+	if err := r.RemoteClients[clusterID].Create(context.TODO(), remoteNamespace); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			if errGet := r.RemoteClients[clusterID].Get(context.TODO(), types.NamespacedName{Name: remoteName}, remoteNamespace); errGet == nil {
+				if value, ok := remoteNamespace.Annotations[annotationKey]; ok && value == remoteNamespaceAnnotationValue {
+					klog.Infof("Namespace '%s' already created inside remote cluster: '%s'", remoteNamespace.Name, clusterID)
+					return nil
+				}
+			}
+		}
+		klog.Error(err)
+		return err
+	}
+
+	klog.Infof("Namespace '%s' created inside remote cluster: '%s'", remoteNamespace.Name, clusterID)
+	return nil
+}
+
+// This function tries to create all the remote namespaces requested in DesiredMapping (NamespaceMap->Spec->DesiredMapping).
+func (r *NamespaceMapReconciler) ensureRemoteNamespacesExistence(nm *mapsv1alpha1.NamespaceMap) int {
+	errorCondition := 0
+	for localName, remoteName := range nm.Spec.DesiredMapping {
+		if err := r.createRemoteNamespace(nm.Labels[liqoconst.RemoteClusterID], remoteName); err != nil {
+			nm.Status.CurrentMapping[localName] = mapsv1alpha1.RemoteNamespaceStatus{
+				RemoteNamespace: remoteName,
+				Phase:           mapsv1alpha1.MappingCreationLoopBackOff,
+			}
+			errorCondition = 1
```

Comment:
If you want to return a counter this should be incremented instead of set to one, otherwise use a boolean variable.

Reply:
I have used a counter because in the calling function "ensureRemoteNamespaces"  I have summed this int value in order to understand if in one of the two called functions there was an error or not. So here I can use a boolean you are right, but in the calling function I need this int value, so I can increment the counter here even if I don't care the number of error that I have, just one is enough...

Reply:
I have incremented the counters at the end

Reply:
> I have used a counter because in the calling function "ensureRemoteNamespaces" I have summed this int value in order to understand if in one of the two called functions there was an error or not. So here I can use a boolean you are right, but in the calling function I need this int value, so I can increment the counter here even if I don't care the number of errors that I have, just one is enough...

I also think that int are a bit strange in this situation. You can obtain the same with the OR operand.

Reply:
I have used boolean with OR in the calling function